### PR TITLE
TAP integration tests

### PR DESCRIPTION
**Pull request by stefwalter**
Orchestrate the integration tests using TAP by producing TAP output. In addition schedule test jobs internally in python.

### CODE REVIEW

**Comment by stefwalter** [comment]
I had to remove RepeatThis for this to work. If we do still need it, then we probably have to make it happen at a different level.

**Comment by mvollmer** [comment]
> I had to remove RepeatThis for this to work. If we do still need it, then we probably have to make it happen at a different level.

A RepeatThis exception does very occasionally happen.  But it's good to remove them all and put them in again as needed.

File: test/check-docker [hunk 1]
```diff
@@ -50,12 +50,7 @@ class TestDocker(MachineCase):
 
     def wait_for_message(self, message):
         b = self.browser
-        # HACK - sometimes container output is missing from the logs
-        #        https://github.com/docker/docker/issues/10617
```
Comment (mvollmer) [inline test/check-docker]: Let's leave a comment here with a link to the bug as a clue why this might be failing.

File: test/check-session [hunk 2]
```diff
@@ -39,10 +39,8 @@ class TestSession(MachineCase):
                 # HACK - sometimes the session gets stuck forever in
                 #        state 'closing' with no processes.
                 #        https://bugs.freedesktop.org/show_bug.cgi?id=89024
-                try:
-                    wait(cond)
-                except:
-                    raise RepeatThis("https://bugs.freedesktop.org/show_bug.cgi?id=89024")
+                # wait(cond)
```
Comment (mvollmer) [inline test/check-session]: The `wait(cond)` is important.  Without RepeatThis, we don't need the `if` at all, we can just do `wait(cond)` always.

**Comment by mvollmer** [comment]
Awesome!

**Comment by mvollmer** [comment]
The output from concurrent tests is interleaved now.  That's makes it significantly harder to read the log.

What about writing test output to one file per test?  We might want to do that anyway when we want to make a nice looking result page.

**Comment by mvollmer** [comment]
There doesn't seem to be much in the log anymore... I'll check what hubbot does with stderr.

**Comment by mvollmer** [comment]
> I'll check what hubbot does with stderr.

Look ok... hmm..

**Comment by mvollmer** [comment]
The failure images are now called like `check-verify-testBasic.png` and we have a lot of tests called `testBasic` so I think they will overwrite each other.

**Comment by mvollmer** [comment]
About the lost output: `check-verify` alone shows the TAP output

```
./check-verify
1..68
# ----------------------------------------------------------------------
# testUsed (check_storage_used.TestStorage)
#
# ----------------------------------------------------------------------
# testBasic (check_pages.TestPages)
#
# ----------------------------------------------------------------------
# testAuthConfig (check_kubernetes.TestConnection)
#
# ----------------------------------------------------------------------
# testFrameReload (check_pages.TestPages)
#
```

but when piping the output through `cat` it doesn't show it anymore:

```
./check-verify | cat
> child frame cockpit1:localhost/playground/test sending data without init
> child frame cockpit1:localhost/playground/test sending data without init
> child frame cockpit1:localhost/playground/test sending data without init
```

**Comment by mvollmer** [comment]
Test results are green although there are failures.

**Comment by dperpeet** [comment]
> The output from concurrent tests is interleaved now. That's makes it significantly harder to read the log.
> 
> What about writing test output to one file per test? We might want to do that anyway when we want to make a nice looking result page.

Yes, definitely! One output file per test would be a lot better. This is not just better for displaying results, but it also makes JOIN easier if we parallelize the tests.

**Comment by stefwalter** [comment]
> Yes, definitely! One output file per test would be a lot better. This is not just better for displaying results, but it also makes JOIN easier if we parallelize the tests.

automake does this and makes debugging really hard. Have you used the *.log files that 'make check' produces? Those are the one file per test files.

I'd like to fix this without creating one file per test.

**Comment by dperpeet** [comment]
> automake does this and makes debugging really hard. Have you used the *.log files that 'make check' produces? Those are the one file per test files.
> 
> I'd like to fix this without creating one file per test.

One file per test is good, we can merge this automatically at the end of our default run to produce a single page. I'd also be happy with something that's easy to split by a script.

**Comment by dperpeet** [comment]
regarding this error:
```
# ----------------------------------------------------------------------
# testIpa (check_realms.TestRealms)
#
not ok 68 testIpa (check_realms.TestRealms)
Traceback (most recent call last):
  File "check-realms", line 40, in testIpa
    ipa.start()
  File "/home/hubbot/hubbot/527dfb156897e32ca7b3576de6615e932a70bbf2_f22_x86-64/test/testvm.py", line 844, in start
    self._start_qemu(maintain=maintain, macaddr=macaddr)
  File "/home/hubbot/hubbot/527dfb156897e32ca7b3576de6615e932a70bbf2_f22_x86-64/test/testvm.py", line 769, in _start_qemu
    macaddr = self._choose_macaddr()
  File "/home/hubbot/hubbot/527dfb156897e32ca7b3576de6615e932a70bbf2_f22_x86-64/test/testvm.py", line 737, in _choose_macaddr
    raise Failure("Couldn't find unused mac address for '%s'" % (self.flavor))
Failure: Couldn't find unused mac address for 'ipa'
Wrote TestRealms-testIpa-FAIL.png
Warning: Permanently added '10.111.125.245' (ECDSA) to the list of known hosts.
Journal database copied to TestRealms-testIpa-10.111.125.245-FAIL.journal
```
There is a potential fix in #2963, but I don't think that's enough. The changes in this pull request parallelize the called tests differently: There are resource races happening here! In this case, both `check-realms` tests want an ipa instance. Yet only one can exist at any given moment. Maybe we can move the two tests into the same class? In the case of `ipa` multiple instances won't work either, because the software in that image requires a specific ip address.

**Comment by stefwalter** [comment]
Should be fixed now.

**Comment by stefwalter** [comment]
Rebased.

File: test/check-docker (hunk 1 above)
Comment (dperpeet) [inline test/check-docker]: I second this

File: test/check-verify [hunk 3]
```diff
@@ -1,77 +1,37 @@
...
+    if name == "check-verify":
+        return None
+    allowed = string.ascii_letters + string.digits + '-_'
+    if not all(c in allowed for c in name):
+        return None
+    return name.replace("-", "_")
 
-echo VERIFICATION PASSED
+# Load all modules that start with check-
+for filename in glob.glob(os.path.join(base, "check-*")):
```
Comment (dperpeet) [inline test/check-verify]: Should we exclude check-example here? It doesn't really make sense to include that...

Maybe we can go further and remove the example? There are lots of simple tests to pick from when learning what they do.

Comment (stefwalter) [inline test/check-verify]: It's a very useful test when diagnosing test and job issues. I can exclude it though.

File: test/check-docker (hunk 1 above)
Comment (stefwalter) [inline test/check-docker]: Fixed.

File: test/check-session (hunk 2 above)
Comment (stefwalter) [inline test/check-session]: Fixed.

**Comment by stefwalter** [comment]
Excluded check-exmaple and  fixed remaining review issues.

**Comment by mvollmer** [comment]
>  fixed remaining review issues.

lgtm.  I wait for r7 results and then merge.

**Pull request closed without merging**